pkg/tests/ossm/deploy_on_infra_test.go

```diff
@@ -0,0 +1,183 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ossm
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	"golang.org/x/exp/slices"
+
+	"github.com/maistra/maistra-test-tool/pkg/util"
+	"github.com/maistra/maistra-test-tool/pkg/util/check/assert"
+	"github.com/maistra/maistra-test-tool/pkg/util/env"
+	"github.com/maistra/maistra-test-tool/pkg/util/oc"
+	"github.com/maistra/maistra-test-tool/pkg/util/pod"
+	"github.com/maistra/maistra-test-tool/pkg/util/retry"
+	"github.com/maistra/maistra-test-tool/pkg/util/shell"
+	"github.com/maistra/maistra-test-tool/pkg/util/test"
+	. "github.com/maistra/maistra-test-tool/pkg/util/test"
+	"github.com/maistra/maistra-test-tool/pkg/util/version"
+)
+
+var workername string
+
+func TestDeployOnInfraNodes(t *testing.T) {
+	NewTest(t).Id("T40").Groups(Full).Run(func(t TestHelper) {
+		t.Log("This test verifies that the OSSM operator and Istio components can be configured to run on infrastructure nodes")
+		if !env.GetSMCPVersion().GreaterThanOrEqualTo(version.SMCP_2_3) {
+			t.T().Skip("Deploy On Infra node is available in SMCP versions v2.3+")
+		}
+		if Smcp.Rosa {
+			t.Skip("Skipping test on ROSA due to lack of permissions")
+		}
+		t.Cleanup(func() {
+			shell.Execute(t,
+				`oc adm taint nodes -l node-role.kubernetes.io/infra node-role.kubernetes.io/infra=reserved:NoSchedule- node-role.kubernetes.io/infra=reserved:NoExecute-`)
+			oc.Label(t, "", "node", workername, "node-role.kubernetes.io/infra-")
+			oc.Label(t, "", "node", workername, "node-role.kubernetes.io-")
+		})
+
+		t.LogStep("Setup: Get a worker node from the cluster that does not have the istio operator installed, label it as infra")
+		workername = pickWorkerNode(t)
+		oc.Label(t, "", "node", workername, "node-role.kubernetes.io/infra=")
+		oc.Label(t, "", "node", workername, "node-role.kubernetes.io=infra")
+		t.Log(fmt.Sprintf("Worker node selected: %s", workername))
+
+		t.LogStep("Taint node and edit the subscription to add the infra node to the node selector")
+		shell.Execute(t,
+			`oc adm taint nodes -l node-role.kubernetes.io/infra node-role.kubernetes.io/infra=reserved:NoSchedule node-role.kubernetes.io/infra=reserved:NoExecute`)
+
+		t.NewSubTest("operator").Run(func(t TestHelper) {
+			t.Log("Verify OSSM Operator is deployed on infra node when configured")
+			t.Log("Reference: https://issues.redhat.com/browse/OSSM-2342")
+			t.Cleanup(func() {
+				shell.Execute(t,
+					`oc patch subscription servicemeshoperator -n openshift-operators --type json -p='[{"op": "remove", "path": "/spec/config/tolerations"}]'`)
+				oc.WaitSMCPReady(t, meshNamespace, smcpName) // Wait for the SMCP to be ready before move to next subtest case (Because untaint)
```

Comment:
This probably doesn't wait at all, because the SMCP will still be ready such a short time after you patch the subscription.

Reply:
I'll delete this because I saw that you add a setup inside each test case, so will check if smcp is ready or not in each test